cmd/frontend/graphqlbackend/graphqlbackend.go

```diff
@@ -730,3 +741,163 @@ func (r *schemaResolver) PhabricatorRepo(ctx context.Context, args *struct {
 func (r *schemaResolver) CurrentUser(ctx context.Context) (*UserResolver, error) {
 	return CurrentUser(ctx)
 }
+
+func (r *schemaResolver) AffiliatedRepositories(ctx context.Context, args *struct {
+	User     graphql.ID
+	CodeHost *graphql.ID
+	Page     *int32
+	Query    *string
+}) (*codeHostRepositoryConnectionResolver, error) {
+	userID, err := UnmarshalUserID(args.User)
+	if err != nil {
+		return nil, err
+	}
+	// 🚨 SECURITY: make sure the user is either site admin or the same user being requested
+	if err := backend.CheckSiteAdminOrSameUser(ctx, userID); err != nil {
+		return nil, err
+	}
+	var codeHost int64
+	if args.CodeHost != nil {
+		codeHost, err = unmarshalExternalServiceID(*args.CodeHost)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &codeHostRepositoryConnectionResolver{
+		userID:   userID,
+		codeHost: codeHost,
+		page: func() int32 {
+			if args.Page != nil {
+				return *args.Page
+			}
+			return 0
+		}(),
+		query: func() string {
+			if args.Query != nil {
+				return *args.Query
+			}
+			return ""
+		}(),
+	}, nil
+}
+
+type codeHostRepositoryConnectionResolver struct {
+	userID   int32
+	codeHost int64
+	page     int32
+	query    string
+
+	once  sync.Once
+	nodes []*codeHostRepositoryResolver
+	err   error
+}
+
+func (r *codeHostRepositoryConnectionResolver) Nodes(ctx context.Context) ([]*codeHostRepositoryResolver, error) {
+	r.once.Do(func() {
+		var (
+			svcs []*types.ExternalService
+			err  error
+		)
+		// get all external services for user, or for the specified external service
+		if r.codeHost == 0 {
+			svcs, err = db.ExternalServices.List(ctx, db.ExternalServicesListOptions{NamespaceUserID: r.userID})
+			if err != nil {
+				r.err = err
+				return
+			}
+		} else {
+			svc, err := db.ExternalServices.GetByID(ctx, r.codeHost)
+			if err != nil {
+				r.err = err
+				return
+			}
+			// 🚨 SECURITY: if the user doesn't own this service, check they're site admin
+			if svc.NamespaceUserID != r.userID {
+				backend.CheckUserIsSiteAdmin(ctx, r.userID)
+			}
+			svcs = []*types.ExternalService{svc}
+		}
+		// get Source for all external services
+		var (
+			results = make(chan []types.CodeHostRepository, 1024)
+			g       = errgroup.Group{}
+			done    = make(chan struct{})
+		)
+		// collect all results in a goroutine
+		go func() {
+			defer close(done)
+			r.nodes = []*codeHostRepositoryResolver{}
+			for repos := range results {
+				for _, repo := range repos {
+					repo := repo
+					r.nodes = append(r.nodes, &codeHostRepositoryResolver{
+						repo: &repo,
+					})
+				}
+			}
+		}()
+		for _, svc := range svcs {
+			src, err := repos.NewSource(svc, cf)
+			if err != nil {
+				r.err = err
+				return
+			}
+			if af, ok := src.(repos.AffiliatedRepositorySource); ok {
+				g.Go(func() error {
+					repos, err := af.AffiliatedRepositories(ctx)
+					if err != nil {
+						return err
+					}
+					results <- repos
```

Comment:
Together with `errgroup.WithContext`, this makes it so we close all worker go routines as soon as any of them fails, or there's a timeout in the parent context.

```suggestion
					select {
					case results <- repos:
					case <-ctx.Done():
				        return ctx.Err()
					}
```